File: coding-test-1/practice-problems/solutions/07-most-letters.rb
# Write a method that takes a string in and returns true if the letter
# "z" appears within three letters **after** an "a". You may assume
# that the string contains only lowercase letters.
#
# Difficulty: medium.

def nearby_az(string)
  idx1 = 0
  while idx1 < string.length
    if string[idx1] != "a"
      idx1 += 1
      next
    end

    idx2 = idx1 + 1
    while (idx2 < string.length) && (idx2 <= idx1 + 3)
      if string[idx2] == "z"
        return true
      end

      idx2 += 1
    end

    idx1 += 1
  end

  return false
end

puts("nearby_az(\"baz\") == true: #{nearby_az("baz") == true}")
puts("nearby_az(\"abz\") == true: #{nearby_az("abz") == true}")
puts("nearby_az(\"abcz\") == true: #{nearby_az("abcz") == true}")
puts("nearby_az(\"a\") == false: #{nearby_az("a") == false}")
puts("nearby_az(\"z\") == false: #{nearby_az("z") == false}")
puts("nearby_az(\"za\") == false: #{nearby_az("za") == false}")

File: coding-test-1/practice-problems/solutions-python/07-most-letters.py
# Write a method that takes a string in and returns true if the letter
# "z" appears within three letters **after** an "a". You may assume
# that the string contains only lowercase letters.
#
# Difficulty: medium.
import unittest


def nearby_az(string):
    string = list(string)

    if "a" not in string:
        return False

    else:
        az_found = False
        for index, letter in enumerate(string):
            if letter == "a" and "z" in string[index:index+4]:
                az_found = True
        return az_found

# These are tests to check that your code is working. After writing
# your solution, they should all run.


class MostLetters(unittest.TestCase):

    def test_baz(self):
        self.assertTrue(nearby_az("baz"))

    def test_abz(self):
        self.assertTrue(nearby_az("abz"))

    def test_abcz(self):
        self.assertTrue(nearby_az("abcz"))

    def test_single_a(self):
        self.assertFalse(nearby_az("a"))

    def test_single_z(self):
        self.assertFalse(nearby_az("z"))

    def test_za(self):
        self.assertFalse(nearby_az("za"))

if __name__ == '__main__':
    unittest.main()